### DISCUSSION TITLE
How to discern nodes based on their order?

### DISCUSSION THREAD

**Discussion by debugloop** [ast-grep/ast-grep#675] [Q&A] [answered]
I've been playing around with using ast-grep for Go's structured logging library slog to tame a growing zoo of different formats in log field names. In addition to predeclaring fields, it also supports variadically adding log fields by invoking `slog.Info("msg", "key", "value", "morekeys", "values")`. I'd like to enforce kebab case on every key, but I can't get it to work. 

The following Playground link demonstrates how to fix the first key and preserve the latter arguments. I'd be happy already if it would fix the last non-conforming key (second to last element in the `argument_list`) instead of the first. That would allow me to fix all bad fields on consecutive runs of ast-grep. My idea has been too add a `$$$PREV_VALUES` before the constrained `$KEY`, but it appears not greedy enough and always matches zero nodes.

Even better would be the option to fix all keys in one go, without iterating executions of ast-grep to get the last non-conforming key, but as I said, I'd be happy with the first already. I've read about `util` rules and the possibility to go recursive, but I'm just not there yet understanding-wise to tackle this, but I feel this might be a way to go...

[Playground Link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImdvIiwicXVlcnkiOiIoJCQkKSIsInJld3JpdGUiOiIiLCJjb25maWciOiItLS1cbmlkOiBzbG9nLWtlYmFiLWR5bmFtaWMtY2FsbFxubWVzc2FnZTogc2xvZyBmaWVsZCBkb2VzIG5vdCB1c2Uga2ViYWItY2FzZVxubGFuZ3VhZ2U6IGdvXG5ydWxlOlxuICBwYXR0ZXJuOlxuICAgIGNvbnRleHQ6ICd7IHNsb2cuJEZVTkMoJE1TRywgJEtFWSwgJCQkVkFMVUVTKSB9J1xuICAgIHNlbGVjdG9yOiBjYWxsX2V4cHJlc3Npb25cbmNvbnN0cmFpbnRzOlxuICBGVU5DOlxuICAgIHJlZ2V4OiAnXihEZWJ1Z3xJbmZvfFdhcm58RXJyb3IpJCdcbiAgS0VZOlxuICAgIHJlZ2V4OiAnXlwiLio/W0EtWi5fIC9dLio/XCIkJ1xudHJhbnNmb3JtOlxuICBCQVJFX0tFWTpcbiAgICBzdWJzdHJpbmc6XG4gICAgICBzb3VyY2U6ICRLRVlcbiAgICAgIHN0YXJ0Q2hhcjogMVxuICAgICAgZW5kQ2hhcjogLTFcbiAgS0VCQUJfS0VZOlxuICAgIGNvbnZlcnQ6XG4gICAgICBzb3VyY2U6ICRCQVJFX0tFWVxuICAgICAgdG9DYXNlOiBrZWJhYkNhc2VcbmZpeDogJ3Nsb2cuJEZVTkMoJE1TRywgXCIkS0VCQUJfS0VZXCIsICQkJFZBTFVFUyknIiwic291cmNlIjoiZnVuYyBtYWluKCkge1xuICBzbG9nLldhcm4oXCJUaGlzIGlzIHRoZSBtZXNzYWdlIVwiLCBcIkZpZWxkS2V5T25lXCIsIFwiZmllbGQtY29udGVudC1vbmVcIiwgXCJmaWVsZC1rZXlfdHdvXCIsIDMuMTQpXG4gIHNsb2cuV2FybihcIlRoaXMgaXMgdGhlIG1lc3NhZ2UhXCIsIFwiZmllbGQta2V5LW9uZVwiLCBcImZpZWxkLWNvbnRlbnQtb25lXCIsIFwiZmllbGQta2V5X3R3b1wiLCAzLjE0KVxufSJ9)

On a similar note: The amount of args should always be odd 🤔

**Comment by HerringtonDarkholme** [comment]
Hi @debugloop 

Thanks for you valuable feedback! I appreciate all your questions and reporting and they gave me many insights on how users are using ast-grep.

For the use case in this post, I think it may be impossible to fix multiple nodes or nodes. At the moment, ast-grep fixing can only fix one target ast node at a time. And telling node's order is not possible at the moment either. :( 

I didn't support them because it may be too complex for the rule YAML. 

I can try using YAML rule to implement this change. But I would recommend to use programmatic API. 

Sadly, currently python API is still under development. https://github.com/ast-grep/ast-grep/issues/672

NodeJS API has not supported the Python language yet.

**Reply by debugloop** [reply]
Thanks for still trying, I'd rather stick to YAML as I intend to use it as a pipeline component where others can add there rule in a simple and straightforward way, without modifying/extending the pipeline by further calls or executables.

**Comment by HerringtonDarkholme** [comment]
Since util rule is recursive, we can always roll out order-specific rules by hand. (alas, [peano-number](https://wiki.haskell.org/Peano_numbers))

In this case, we can define the key node as either
1) follows the first argument `message`
2) follows a `value` argument which itself follows another `key` node.

```yaml
id: slog-kebab-dynamic-call
message: slog field does not use kebab-case
language: go
utils:
  inside-slog:
    pattern:
      context: '{ slog.$FUNC($$$) }'
      selector: argument_list
  first-key:
    not:
      follows:
        pattern: $_
        stopBy:
          pattern: $_
  at-key:
    kind: interpreted_string_literal
    follows:
      stopBy:
        pattern: $_
      pattern: $_
      any:
        - matches: first-key
        - follows:
            matches: at-key
            stopBy:
              pattern: $_
rule:
  matches: at-key
  pattern: $KEY
  inside:
    matches: inside-slog
constraints:
  FUNC:
    regex: '^(Debug|Info|Warn|Error)$'
  KEY:
    regex: '^".*?[A-Z._ /].*?"$'
transform:
  BARE_KEY:
    substring:
      source: $KEY
      startChar: 1
      endChar: -1
  KEBAB_KEY:
    convert:
      source: $BARE_KEY
      toCase: kebabCase
fix: '"$KEBAB_KEY"'

```

<img width="668" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/9a690f29-9e09-4013-81c9-b2e0d45ebb2c">



[Rule Playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImdvIiwicXVlcnkiOiIoJCQkKSIsInJld3JpdGUiOiIiLCJjb25maWciOiJpZDogc2xvZy1rZWJhYi1keW5hbWljLWNhbGxcbm1lc3NhZ2U6IHNsb2cgZmllbGQgZG9lcyBub3QgdXNlIGtlYmFiLWNhc2Vcbmxhbmd1YWdlOiBnb1xudXRpbHM6XG4gIGluc2lkZS1zbG9nOlxuICAgIHBhdHRlcm46XG4gICAgICBjb250ZXh0OiAneyBzbG9nLiRGVU5DKCQkJCkgfSdcbiAgICAgIHNlbGVjdG9yOiBhcmd1bWVudF9saXN0XG4gIGZpcnN0LWtleTpcbiAgICBub3Q6XG4gICAgICBmb2xsb3dzOlxuICAgICAgICBwYXR0ZXJuOiAkX1xuICAgICAgICBzdG9wQnk6XG4gICAgICAgICAgcGF0dGVybjogJF9cbiAgYXQta2V5OlxuICAgIGtpbmQ6IGludGVycHJldGVkX3N0cmluZ19saXRlcmFsXG4gICAgZm9sbG93czpcbiAgICAgIHN0b3BCeTpcbiAgICAgICAgcGF0dGVybjogJF9cbiAgICAgIHBhdHRlcm46ICRfXG4gICAgICBhbnk6XG4gICAgICAgIC0gbWF0Y2hlczogZmlyc3Qta2V5XG4gICAgICAgIC0gZm9sbG93czpcbiAgICAgICAgICAgIG1hdGNoZXM6IGF0LWtleVxuICAgICAgICAgICAgc3RvcEJ5OlxuICAgICAgICAgICAgICBwYXR0ZXJuOiAkX1xucnVsZTpcbiAgbWF0Y2hlczogYXQta2V5XG4gIHBhdHRlcm46ICRLRVlcbiAgaW5zaWRlOlxuICAgIG1hdGNoZXM6IGluc2lkZS1zbG9nXG5jb25zdHJhaW50czpcbiAgRlVOQzpcbiAgICByZWdleDogJ14oRGVidWd8SW5mb3xXYXJufEVycm9yKSQnXG4gIEtFWTpcbiAgICByZWdleDogJ15cIi4qP1tBLVouXyAvXS4qP1wiJCdcbnRyYW5zZm9ybTpcbiAgQkFSRV9LRVk6XG4gICAgc3Vic3RyaW5nOlxuICAgICAgc291cmNlOiAkS0VZXG4gICAgICBzdGFydENoYXI6IDFcbiAgICAgIGVuZENoYXI6IC0xXG4gIEtFQkFCX0tFWTpcbiAgICBjb252ZXJ0OlxuICAgICAgc291cmNlOiAkQkFSRV9LRVlcbiAgICAgIHRvQ2FzZToga2ViYWJDYXNlXG5maXg6ICdcIiRLRUJBQl9LRVlcIidcbiIsInNvdXJjZSI6ImZ1bmMgbWFpbigpIHtcbiAgc2xvZy5XYXJuKCBcIlRoaXMgaXMgdGhlIG1lc3NhZ2UhXCIsIFwiRmllbGRLZXlPbmVcIiwgXCJmaWVsZC1jb250ZW50LW9uZVwiLCBcImZpZWxkLWtleV90d29cIiwgXCJzZGZzZGZcIiwgXCJzZGZzZGZcIiwgXCJzZGRmc2Rmc2RcIilcbiAgc2xvZy5XYXJuKFwiVGhpcyBpcyB0aGUgbWVzc2FnZSFcIiwgXCJmaWVsZC1rZXktb25lXCIsIFwiZmllbGQtY29udGVudF9vbmVcIiwgXCJmaWVsZC1rZXlfdHdvXCIsIDMuMTQpXG59In0=)

I would think about how to express the "quantity/order" concept in YAML rule. But currently this is a workaround.

**Reply by debugloop** [reply]
Thanks so much, that get's me a lot farther in my quest 🙏🏻 I kinda don't understand why the outermost pattern (inside-slog) does not really restrict the results any more, and also the FUNC constraint seems to be ineffective. I've tried pulling the (inside-slog) match into the top level pattern, but it seems the ineffectiveness of the constraints is not related to the invoke through a `match`... See [here](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImdvIiwicXVlcnkiOiIoJCQkKSIsInJld3JpdGUiOiIiLCJjb25maWciOiJpZDogc2xvZy1rZWJhYi1keW5hbWljLWNhbGxcbm1lc3NhZ2U6IHNsb2cgZmllbGQgZG9lcyBub3QgdXNlIGtlYmFiLWNhc2Vcbmxhbmd1YWdlOiBnb1xudXRpbHM6XG4gIGluc2lkZS1zbG9nOlxuICAgIHBhdHRlcm46XG4gICAgICBjb250ZXh0OiAneyBzbG9nLiRGVU5DKCQkJCkgfSdcbiAgICAgIHNlbGVjdG9yOiBhcmd1bWVudF9saXN0XG4gIGZpcnN0LWtleTpcbiAgICBub3Q6XG4gICAgICBmb2xsb3dzOlxuICAgICAgICBwYXR0ZXJuOiAkX1xuICAgICAgICBzdG9wQnk6XG4gICAgICAgICAgcGF0dGVybjogJF9cbiAgYXQta2V5OlxuICAgIGtpbmQ6IGludGVycHJldGVkX3N0cmluZ19saXRlcmFsXG4gICAgZm9sbG93czpcbiAgICAgIHN0b3BCeTpcbiAgICAgICAgcGF0dGVybjogJF9cbiAgICAgIHBhdHRlcm46ICRfXG4gICAgICBhbnk6XG4gICAgICAgIC0gbWF0Y2hlczogZmlyc3Qta2V5XG4gICAgICAgIC0gZm9sbG93czpcbiAgICAgICAgICAgIG1hdGNoZXM6IGF0LWtleVxuICAgICAgICAgICAgc3RvcEJ5OlxuICAgICAgICAgICAgICBwYXR0ZXJuOiAkX1xucnVsZTpcbiAgbWF0Y2hlczogYXQta2V5XG4gIHBhdHRlcm46ICRLRVlcbiAgaW5zaWRlOlxuICAgIG1hdGNoZXM6IGluc2lkZS1zbG9nXG5jb25zdHJhaW50czpcbiAgRlVOQzpcbiAgICByZWdleDogJ14oRGVidWd8SW5mb3xXYXJufEVycm9yKSQnXG4gIEtFWTpcbiAgICByZWdleDogJ15cIi4qP1tBLVouXyAvXS4qP1wiJCdcbnRyYW5zZm9ybTpcbiAgQkFSRV9LRVk6XG4gICAgc3Vic3RyaW5nOlxuICAgICAgc291cmNlOiAkS0VZXG4gICAgICBzdGFydENoYXI6IDFcbiAgICAgIGVuZENoYXI6IC0xXG4gIEtFQkFCX0tFWTpcbiAgICBjb252ZXJ0OlxuICAgICAgc291cmNlOiAkQkFSRV9LRVlcbiAgICAgIHRvQ2FzZToga2ViYWJDYXNlXG5maXg6ICdcIiRLRUJBQl9LRVlcIidcbiIsInNvdXJjZSI6ImZ1bmMgbWFpbigpIHtcbiAgc2xvZy5XYXJuKCBcIlRoaXMgaXMgdGhlIG1lc3NhZ2UhXCIsIFwiRmllbGRLZXlPbmVcIiwgXCJmaWVsZC1jb250ZW50LW9uZVwiLCBcImZpZWxkLWtleV90d29cIiwgXCJzZGZzZGZcIiwgXCJzZGZzZGZcIiwgXCJzZGRmc2Rmc2RcIilcbiAgc2xvZy5XYXJuKFwiVGhpcyBpcyB0aGUgbWVzc2FnZSFcIiwgXCJmaWVsZC1rZXktb25lXCIsIFwiZmllbGQtY29udGVudF9vbmVcIiwgXCJmaWVsZC1rZXlfdHdvXCIsIDMuMTQpXG4gIG90aGVyLkxpYkZ1bmN0aW9uKFwiVGhpcyBpcyB0aGUgbWVzc2FnZSFcIiwgXCJmaWVsZC1rZXktb25lXCIsIFwiZmllbGQtY29udGVudF9vbmVcIiwgXCJmaWVsZC1rZXlfdHdvXCIsIDMuMTQpICBcbn0ifQ==).

**Reply by HerringtonDarkholme** [reply, marked as answer]
I forgot to update the selector field! I changed it to other things.

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImdvIiwicXVlcnkiOiIoJCQkKSIsInJld3JpdGUiOiIiLCJjb25maWciOiJpZDogc2xvZy1rZWJhYi1keW5hbWljLWNhbGxcbm1lc3NhZ2U6IHNsb2cgZmllbGQgZG9lcyBub3QgdXNlIGtlYmFiLWNhc2Vcbmxhbmd1YWdlOiBnb1xudXRpbHM6XG4gIGluc2lkZS1zbG9nOlxuICAgIHBhdHRlcm46XG4gICAgICBjb250ZXh0OiAneyBzbG9nLiRGVU5DKCQkJCkgfSdcbiAgICAgIHNlbGVjdG9yOiBjYWxsX2V4cHJlc3Npb25cbiAgZmlyc3Qta2V5OlxuICAgIG5vdDpcbiAgICAgIGZvbGxvd3M6XG4gICAgICAgIHBhdHRlcm46ICRfXG4gICAgICAgIHN0b3BCeTpcbiAgICAgICAgICBwYXR0ZXJuOiAkX1xuICBhdC1rZXk6XG4gICAga2luZDogaW50ZXJwcmV0ZWRfc3RyaW5nX2xpdGVyYWxcbiAgICBmb2xsb3dzOlxuICAgICAgc3RvcEJ5OlxuICAgICAgICBwYXR0ZXJuOiAkX1xuICAgICAgcGF0dGVybjogJF9cbiAgICAgIGFueTpcbiAgICAgICAgLSBtYXRjaGVzOiBmaXJzdC1rZXlcbiAgICAgICAgLSBmb2xsb3dzOlxuICAgICAgICAgICAgbWF0Y2hlczogYXQta2V5XG4gICAgICAgICAgICBzdG9wQnk6XG4gICAgICAgICAgICAgIHBhdHRlcm46ICRfXG5ydWxlOlxuICBtYXRjaGVzOiBhdC1rZXlcbiAgcGF0dGVybjogJEtFWVxuICBpbnNpZGU6XG4gICAgbWF0Y2hlczogaW5zaWRlLXNsb2dcbiAgICBzdG9wQnk6XG4gICAgICBraW5kOiBjYWxsX2V4cHJlc3Npb25cbmNvbnN0cmFpbnRzOlxuICBGVU5DOlxuICAgIHJlZ2V4OiAnXihEZWJ1Z3xJbmZvfFdhcm58RXJyb3IpJCdcbiAgS0VZOlxuICAgIHJlZ2V4OiAnXlwiLio/W0EtWi5fIC9dLio/XCIkJ1xudHJhbnNmb3JtOlxuICBCQVJFX0tFWTpcbiAgICBzdWJzdHJpbmc6XG4gICAgICBzb3VyY2U6ICRLRVlcbiAgICAgIHN0YXJ0Q2hhcjogMVxuICAgICAgZW5kQ2hhcjogLTFcbiAgS0VCQUJfS0VZOlxuICAgIGNvbnZlcnQ6XG4gICAgICBzb3VyY2U6ICRCQVJFX0tFWVxuICAgICAgdG9DYXNlOiBrZWJhYkNhc2VcbmZpeDogJ1wiJEtFQkFCX0tFWVwiJ1xuIiwic291cmNlIjoiZnVuYyBtYWluKCkge1xuICBzbG9nLldhcm4oIFwiVGhpcyBpcyB0aGUgbWVzc2FnZSFcIiwgXCJGaWVsZEtleU9uZVwiLCBcImZpZWxkLWNvbnRlbnQtb25lXCIsIFwiZmllbGQta2V5X3R3b1wiLCBcInNkZnNkZlwiLCBcInNkZnNkZlwiLCBcInNkZGZzZGZzZFwiKVxuICBzbG9nLldhcm4oXCJUaGlzIGlzIHRoZSBtZXNzYWdlIVwiLCBcImZpZWxkLWtleS1vbmVcIiwgXCJmaWVsZC1jb250ZW50X29uZVwiLCBcImZpZWxkLWtleV90d29cIiwgMy4xNClcbiAgb3RoZXIuTGliRnVuY3Rpb24oXCJUaGlzIGlzIHRoZSBtZXNzYWdlIVwiLCBcImZpZWxkLWtleS1vbmVcIiwgXCJmaWVsZC1jb250ZW50X29uZVwiLCBcImZpZWxkLWtleV90d29cIiwgMy4xNCkgIFxufSJ9

**Reply by debugloop** [reply]
Thank you again, got my head around it and managed to adapt it to work on `WarnContext(ctx, ...` calls as well 🎉 I got one more question though: Performance wise this seems way worse than just a plain lookup for `slog.$FUNC`, why is that? Does the recursion run on all functions regardless of the outer match? I.e. the `inside` part is checked only after the initial `match` is through?

Rewriting to the following does not seem to change things, even though I can specify order this way:

```yaml
rule:
  all:
    - inside:
        matches: inside-slog
        stopBy:
          kind: call_expression
    - matches: at-key
    - pattern: $KEY
```

**Reply by HerringtonDarkholme** [reply]
Using `all` can specify rule order so it is generally faster than using object style rule. (Object style rule is for more concise rule)

So in this case, it will first run the recursion rule. Using `all` can let us choose which rule to run first!

Haha this is an advanced usage of `all` rule (currently it is not documented). Glad you find it!

**Reply by debugloop** [reply]
Yes I thought it would be this way, but it's not really different I'm afraid. It seems like the fixed order does not have an effect sadly...
